### PR TITLE
fix: retrying config loading errors that lead to new errors freezes initialization

### PR DESCRIPTION
**Pull request by ryanthemanuel**
- Closes UNIFY-1467

### User facing changelog

When the app is in an error state, and the user clicks "Try again" and the result is another error, the new error is now displayed instead of getting stuck in the initialization state.

### Additional details

Previously, we were not emitting to the launchpad that we had begun initialization again on refresh. This was causing a race condition where the error state wasn't properly being updated on the client. Going forward, we are going to ensure that we are emitting to the launchpad right after initialization and returning immediately rather than waiting for refreshing to be done. This will allow the client UI to be properly updated. In addition, in situations like migration and project, this will make sure that the user receives some feedback after completing the workflows.

While working through this issue, I noticed a few places that could be cleaned up as well. There was some dead code and some code that seemed to more closely belong in other data sources or actions.

### PR Tasks

- [X] Have tests been added/updated?
- [na] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [na] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

**Review by lmiller1990 — COMMENTED** [review]
I left some comments and thoughts. We definitely need better patterns around testing the Lifecycle and Config manager... I think the e2e test you added are pretty good, but the value of the unit tests is debatable. They are similar to `server` in the sense they stub a lot and test implementation details, as opposed to the actual public facing behavior.  

I think we need a kind of test harness of function to spin up an entire config process and make assertions on things, as opposed to relying on stubbing methods out all over the place. The tests will be slower, but they'll be more similar to what the actual code does in production, and give us more confidence.

**Comment by ryanthemanuel** [comment]
@lmiller1990 yeah I think the unit tests here are worth discussing. I originally wasn't going to add one here as my changes are covered by e2e tests; however, based on our discussion in the retro today, I thought I'd throw one out there as a discussion point.

File: packages/data-context/src/actions/MigrationActions.ts [hunk 1]
```diff
@@ -293,7 +293,7 @@ export class MigrationActions {
 
   async finishReconfigurationWizard () {
```
[the remark this answers is not in the document]
Comment (ryanthemanuel) [reply packages/data-context/src/actions/MigrationActions.ts]: Resolved here: [cypress-io/cypress@`53aaf66` (#21176)](https://github.com/cypress-io/cypress/pull/21176/commits/53aaf6669e2943d275e88a5ab8f75b904719a4ff)

File: packages/data-context/src/data/ProjectLifecycleManager.ts [hunk 2]
```diff
@@ -309,24 +283,29 @@ export class ProjectLifecycleManager {
...
+  /**
+   * Kicks off the refresh of the configuration lifecycle and returns immediately. Internally, it handles all errors on behalf of the callers.
+   * These errors will be emitted to the client
+   */
+  refreshLifecycle (): void {
+    this.refreshLifeCycleAsync().catch(this.onLoadError)
+  }
 
-    if (this.readyToInitialize(this._projectRoot)) {
+  private async refreshLifeCycleAsync (): Promise<void> {
```
[the remark this answers is not in the document]
Comment (ryanthemanuel) [reply packages/data-context/src/data/ProjectLifecycleManager.ts]: Resolved here: [cypress-io/cypress@`53aaf66` (#21176)](https://github.com/cypress-io/cypress/pull/21176/commits/53aaf6669e2943d275e88a5ab8f75b904719a4ff)

File: packages/data-context/src/data/ProjectLifecycleManager.ts [hunk 3]
```diff
@@ -309,24 +283,29 @@ export class ProjectLifecycleManager {
...
 
-  async refreshLifecycle () {
-    assert(this._projectRoot, 'Cannot reload config without a project root')
-    assert(this._configManager, 'Cannot reload config without a config manager')
+  /**
+   * Kicks off the refresh of the configuration lifecycle and returns immediately. Internally, it handles all errors on behalf of the callers.
+   * These errors will be emitted to the client
+   */
+  refreshLifecycle (): void {
+    this.refreshLifeCycleAsync().catch(this.onLoadError)
```
[the remark this answers is not in the document]
Comment (ryanthemanuel) [reply packages/data-context/src/data/ProjectLifecycleManager.ts]: > Have we tried this?

This was how the code worked previously and I originally thought it was necessary to change it to fully resolve things. However, I worked through this again and was able to get it working with just the one method. I agree it greatly simplifies things.

File: packages/data-context/src/data/ProjectLifecycleManager.ts [hunk 4]
```diff
@@ -309,24 +283,29 @@ export class ProjectLifecycleManager {
...
+  }
 
-    if (this.readyToInitialize(this._projectRoot)) {
+  private async refreshLifeCycleAsync (): Promise<void> {
+    if (this._projectRoot && this._configManager && this.readyToInitialize(this._projectRoot)) {
       this._configManager.resetLoadingState()
+
+      // Emit here so that the user gets the impression that we're loading rather than waiting for a full refresh of the config for an update
```
[the remark this answers is not in the document]
Comment (ryanthemanuel) [reply packages/data-context/src/data/ProjectLifecycleManager.ts]: Resolved here: [cypress-io/cypress@53aaf66 (#21176)](https://github.com/cypress-io/cypress/pull/21176/commits/53aaf6669e2943d275e88a5ab8f75b904719a4ff)

File: packages/data-context/src/data/ProjectLifecycleManager.ts [hunk 5]
```diff
@@ -268,7 +242,7 @@ export class ProjectLifecycleManager {
 
         this._pendingInitialize?.resolve(finalConfig)
       },
-      refreshLifecycle: async () => this.refreshLifecycle(),
+      refreshLifecycle: async () => await this.refreshLifecycle(),
```
Comment (ryanthemanuel) [inline packages/data-context/src/data/ProjectLifecycleManager.ts:245]:
```suggestion
      refreshLifecycle: async () => this.refreshLifecycle(),
```

File: packages/data-context/test/unit/data/ProjectLifecycleManager.spec.ts [hunk 6]
```diff
@@ -1,13 +1,18 @@
...
+
+chai.use(sinonChai)
 
 describe('ProjectLifecycleManager', () => {
   let ctx: DataContext
+  let lifecycleManager: any
 
   beforeEach(() => {
     ctx = createTestDataContext('open')
+    lifecycleManager = ctx.lifecycleManager as any
```
[the remark this answers is not in the document]
Comment (ryanthemanuel) [reply packages/data-context/test/unit/data/ProjectLifecycleManager.spec.ts]: I reverted these changes: [`bc08833` (#21176)](https://github.com/cypress-io/cypress/pull/21176/commits/bc088331d8b974f6effec6d045e06c1e236fdd91). I definitely agree with you here.

File: packages/data-context/src/data/ProjectLifecycleManager.ts [hunk 7]
```diff
@@ -579,7 +546,7 @@ export class ProjectLifecycleManager {
     const configFile = this.ctx.modeOptions.configFile
     const metaState: ProjectMetaState = {
       ...PROJECT_META_STATE,
-      hasLegacyCypressJson: fs.existsSync(this._pathToFile(this.legacyConfigFile)),
+      hasLegacyCypressJson: fs.existsSync(this._pathToFile(this.ctx.migration.legacyConfigFile)),
```
[the remark this answers is not in the document]
Comment (ryanthemanuel) [reply packages/data-context/src/data/ProjectLifecycleManager.ts]: Resolved here: [`bc08833` (#21176)](https://github.com/cypress-io/cypress/pull/21176/commits/bc088331d8b974f6effec6d045e06c1e236fdd91)

File: packages/data-context/test/unit/data/ProjectLifecycleManager.spec.ts [hunk 8]
```diff
@@ -1,13 +1,18 @@
-import { expect } from 'chai'
+import chai, { expect } from 'chai'
 import type { DataContext } from '../../../src'
 import { createTestDataContext } from '../helper'
 import sinon from 'sinon'
+import sinonChai from 'sinon-chai'
+
+chai.use(sinonChai)
```
[the remark this answers is not in the document]
Comment (ryanthemanuel) [reply packages/data-context/test/unit/data/ProjectLifecycleManager.spec.ts]: With these changes, this is no longer necessary here: [`bc08833` (#21176)](https://github.com/cypress-io/cypress/pull/21176/commits/bc088331d8b974f6effec6d045e06c1e236fdd91)

File: packages/data-context/test/unit/data/ProjectLifecycleManager.spec.ts [hunk 9]
```diff
@@ -47,17 +52,110 @@ describe('ProjectLifecycleManager', () => {
...
+        expect(refreshLifecycleAsyncStub).to.have.been.called
+        done()
+      })
+
+      lifecycleManager.refreshLifecycle()
+    })
+  })
+
+  context('#refreshLifeCycleAsync', () => {
+    it('resets the loading state, notifies launchpad, initializes config, and loads the testing type if ready and configured', async () => {
```
[the remark this answers is not in the document]
Comment (ryanthemanuel) [reply packages/data-context/test/unit/data/ProjectLifecycleManager.spec.ts]: I went ahead and removed the tests (it felt kind of weird adding them). If others want to weigh in on this we can continue the discussion, but for now, the e2e test I added will cover the fix here.

File: packages/data-context/src/data/ProjectLifecycleManager.ts [hunk 10]
```diff
@@ -115,18 +115,6 @@ export class ProjectLifecycleManager {
     return Object.freeze(this._projectMetaState)
   }
 
-  get legacyJsonPath () {
```
Comment (tgriesser) [inline packages/data-context/src/data/ProjectLifecycleManager.ts:118]: Nice cleanup here

File: packages/graphql/src/schemaTypes/objectTypes/gql-CurrentProject.ts [hunk 11]
```diff
@@ -102,7 +102,7 @@ export const CurrentProject = objectType({
     t.boolean('needsLegacyConfigMigration', {
       description: 'Whether the project needs to be migrated before proceeding',
       resolve (source, args, ctx) {
-        return ctx.lifecycleManager.needsCypressJsonMigration()
+        return ctx.migration.needsCypressJsonMigration()
```
Comment (tgriesser) [inline packages/graphql/src/schemaTypes/objectTypes/gql-CurrentProject.ts:105]: 👍

**Review by tgriesser — APPROVED** [review]
One minor comment, not a blocker though

**Review by tbiethman — APPROVED** [review]

**Comment by lmiller1990** [comment]
@ryanthemanuel cool, let's merge this for now, we can continue to discuss and how best to test this stuff moving forward; I think we need to improve the infrastructure in data-context specifically to let us write better tests.

**Pull request merged**